File: 02_Two_Pointers/01_Easy/Python/validPalindrome.py
'''
A phrase is a palindrome if, after converting all uppercase letters into 
lowercase letters and removing all non-alphanumeric characters, it reads 
the same forward and backward. Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
'''

def isValid(s):
    chars = [c.lower() for c in s if c.isalnum()]
    l, r = 0, len(chars) - 1

    while l < r:
        if chars[l] != chars[r]:
            return False
        l += 1 
        r -= 1
    
    return True 

'''
Pretty straightforward. Check from out to in. 

Time | Space: O(n) | O(n)
'''

File: 02_Two_Pointers/01_Easy/C++/ValidPalindrome.cpp
// A phrase is a palindrome if, after converting all uppercase letters
// into lowercase letters and removing all non-alphanumeric characters,
// it reads the same forward and backward. Alphanumeric characters include
// letters and numbers.

// Given a string s, return true if it is a palindrome, or false otherwise.

#include <string>
#include <cctype>

class Solution {
    public:
        bool isPalindrome(std::string s){
            int i = 0; 
            int j = s.size() - 1;

            while (i < j){
                while (!std::isalnum(s[i]) && i < j) i++;
                while (!std::isalnum(s[j]) && i < j) j--;

                if (std::tolower(s[i]) != std::tolower(s[j])) return false;
                i++;
                j--;
            }

            return true;
        };
};

// From the outside in, check every valid character. If unequal, return false.
// Time | Space : O(n) | O(1)
